src/typechecker.erl

```diff
@@ -4227,14 +4250,15 @@ add_type_pat(CONS = {cons, P, PH, PT}, ListTy, Env, VEnv) ->
             TailTy = normalize(type(union, [ListTy, type(nil)]), Env),
             {_TailPatTy, _TauUBound, VEnv3, Cs} = add_type_pat(PT, TailTy, Env, VEnv2),
             NonEmptyTy = rewrite_list_to_nonempty_list(ListTy),
-            {NonEmptyTy, NonEmptyTy, VEnv3, Cs};
+            {type(none), NonEmptyTy, VEnv3, Cs};
```

Comment:
`[_|_]` exhausts nonempty list. Have you saved that for another PR?

Reply:
Hmmm, I have to think it through 🤔

Reply:
I think it's only the pattern `[A | B]` where A and B are free variables (not in VEnv) that works.